### PR TITLE
fix: event engine rewrite 

### PR DESCRIPTION
**Pull request by andreas-karlsson**

### CODE REVIEW

File: src/test/java/com/spotify/confidence/ConfidenceEventSenderIntegrationTest.java [hunk 1]
```diff
@@ -74,10 +74,10 @@ public void testEngineUploadsTriggeredByFlushTimeout() throws IOException, Inter
         "navigate", ConfidenceValue.of(ImmutableMap.of("key", ConfidenceValue.of("size"))));
 
     // wait for the flush timeout to trigger the upload
-    Thread.sleep(200);
+    Thread.sleep(300);
```
Comment (nicklasl) [inline src/test/java/com/spotify/confidence/ConfidenceEventSenderIntegrationTest.java:77]: Restore this to 200 just to make it clear that nothing really changed.

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java [hunk 2]
```diff
@@ -1,162 +1,120 @@
...
-      pendingFlush.cancel(false);
-    }
-    if (flushTimeoutMilliseconds > 0) {
-      pendingFlush =
-          executorService.schedule(this::flush, flushTimeoutMilliseconds, TimeUnit.MILLISECONDS);
-    }
+  @Override
+  public void send(
+      String name, ConfidenceValue.Struct context, Optional<ConfidenceValue.Struct> message) {
+    if (intakeClosed) return;
```
Comment (nicklasl) [inline src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:50]:
```suggestion
    if (intakeClosed || pendingBatches is too big) /* log */ return;
```

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java [hunk 3]
```diff
@@ -1,162 +1,131 @@
...
-            try {
-              closeNow();
-            } catch (IOException e) {
-              // ignore
-            }
-          }
+  private void upload(List<Event> events) {
+    if (events.isEmpty()) return;
+    final CompletableFuture<Boolean> batchUploaded =
+        uploadExecutor.getStageAsync(() -> eventUploader.upload(new EventBatch(events)));
```
Comment (OskarKjellin) [inline src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:98]: What's the purpose of this separate executor? It's just doing non blocking io anyhow, why move it to another thread?

File: src/main/java/com/spotify/confidence/Confidence.java [hunk 4]
```diff
@@ -214,12 +214,8 @@ public Confidence build() {
       final FlagResolverClient flagResolverClient =
           new FlagResolverClientImpl(
               new GrpcFlagResolver(clientSecret, flagResolverManagedChannel));
-      final SystemClock clock = new SystemClock();
-      final GrpcEventUploader uploader =
-          new GrpcEventUploader(clientSecret, clock, DEFAULT_CHANNEL);
```
Comment (OskarKjellin) [inline src/main/java/com/spotify/confidence/Confidence.java:219]: Who owns this channel? It's currently being passed to `GrpcEventUploader` which then closes it in the close method. IMO if passed in it's not owned by `GrpcEventUploader` and the owner should close it

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java (hunk 3 above)
Comment (andreas-karlsson) [reply src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:98]: `uploadExecutor` is the FailSafe executor so it does retries, and needs the ability to schedule retries. Right now it does this on the common pool.

Comment (nicklasl) [reply src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:98]: It's an executor from the `Failsafe` lib which handles retries with backoff internally.

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java [hunk 5]
```diff
@@ -1,162 +1,131 @@
...
     this.clock = clock;
-    this.flushTimeoutMilliseconds = flushTimeoutMilliseconds;
-    writeThread.submit(new WritePoller());
-    uploadThread.submit(new UploadPoller());
+    this.maxBatchSize = maxBatchSize;
+    this.maxFlushInterval = maxFlushInterval;
+    uploadExecutor =
+        Failsafe.with(
+            RetryPolicy.<Boolean>builder()
+                .handleResult(false)
```
Comment (nicklasl) [inline src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:41]: Need to verify is this (upload function returning `false`) is actually a valid reason to retry.

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java (hunk 3 above)
Comment (OskarKjellin) [reply src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:98]: Makes sense, forgot about it being failsafe here

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java [hunk 6]
```diff
@@ -1,162 +1,173 @@
...
-    }
-    if (flushTimeoutMilliseconds > 0) {
-      pendingFlush =
-          executorService.schedule(this::flush, flushTimeoutMilliseconds, TimeUnit.MILLISECONDS);
-    }
+  static Event.Builder event(
+      String name, ConfidenceValue.Struct context, Optional<ConfidenceValue.Struct> message) {
+    return com.spotify.confidence.events.v1.Event.newBuilder()
+        .setEventDefinition(EVENT_NAME_PREFIX + name)
+        .setPayload(
```
Comment (andreas-karlsson) [inline src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:80]: Should we reverse the order of these? Shouldn't message overwrite context?

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java (hunk 5 above)
Comment (andreas-karlsson) [reply src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:58]: So it only returns false on exceptions from grpc. Are there exceptions we don't want to retry?

Comment (andreas-karlsson) [reply src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:58]: Ok, I'm going through the list of possible grpc errors and will make commit a proposal.

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java (hunk 6 above)
Comment (nicklasl) [reply src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:80]: Actually no, with this approach, if the user adds a `context` key in the message, we will overwrite that with the actual context. We want that, not the other way around.

File: src/main/java/com/spotify/confidence/EventSenderEngineImpl.java [hunk 7]
```diff
@@ -1,162 +1,171 @@
...
+      if (event != null) {
+        events.add(event);
+      } else {
+        if (intakeClosed) break;
+        LockSupport.parkUntil(Instant.now().plus(maxFlushInterval).toEpochMilli());
+      }
+      final boolean passedMaxFlushInterval =
+          !maxFlushInterval.isZero()
+              && Duration.between(latestFlushTime, Instant.now()).compareTo(maxFlushInterval) > 0;
+      if (events.size() == maxBatchSize || passedMaxFlushInterval) {
```
Comment (vahidlazio) [inline src/main/java/com/spotify/confidence/EventSenderEngineImpl.java:117]: if I understand the park correctly, if the current event is null, we wait until the `maxFlushInterval` and then here we check if we are passed that time (which in this case will be true) and then we upload events. should we check here as well if events are not empty?

**Review by nicklasl — APPROVED** [review]

**Pull request merged**